Typo checker and fixer added - Closes #771

juhandergod opened the pull request:
## Fixes Issue

Closes #771

## Changes proposed

Hey,
This PR fixes the issue for checking words with typos (#771). In the issue there were different discussions about using some type of NLTK spell checker library. For that use case I added `typo-js-ts` lib with defined dictionary for `US` language. In addition I added some additional logic to remove the non alphabetic characters from each word, and if those characters are removed, then it can check for the typos, and use the first suggestion for that word.

++
I added a testing dependency to the project (called `jest`) to test my new functionaly and cover some older components. 
Updated the Readme, and the build.yml workflow.


### Added:
- Created `typoFixer` component using the `typo-js-ts` dependency for spell checking
- Created unit tests for `typoFixer` component ( `typoFixer.test.ts` )
- Created unit test for `stripSpecialCharacter` component (`stripSpecialCharacter.test.ts` )


- Added dependencies:
  - `typo-js-ts` dependency - Spell checker dependency

- Added `en_US` dictionary for `typo-js-ts` dependency:
  -  (`en_US.aff`, `en_US.dic`)

- Added dev-dependencies for testing with `jest`:
  - `@babel/preset-env`
  - `@babel/preset-typescript`
  - `@jest/globals`
  - `@types/jest`
  - `jest`
  - `ts-jest`


### Changed:
- Updated `package.json` **scripts** section with a new script for running tests
- Updated README.md with `Running the tests` section
- Updated `build.yml` github workflow action with `run tests` step
- Updated dependencies:
  - `mongoose` upgraded to version `6.4.6`
  - `@types/express` upgraded to version `4.17.20`
  -  `@types/node` upgaded to version `14.18.63`
  - `typescript` upgraded to version  `4.9.5`


## Check List (Check all the applicable boxes) <!-- Follow the above conventions to check the box -->

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## Screenshots

![image](https://github.com/EddieHubCommunity/EddieBot/assets/56760600/a7353d1b-a35a-48ba-a4e1-eed8af9a76ab)

## Note to reviewers

Should I update the changelog, or is that generated automatically?

eddiejaoude reviewed (changes requested):
Thank you for the pull request. Some of my thoughts...

✅ great you added tests 🚀 

Would this not cause a problem when people use a word like `picking`?

juhandergod commented:
Hey @eddiejaoude , thanks for your response.

I added one test case to check if it will fix the word picking or not. It will remain the same.

But you are right. This addition can produce false positives. That's the reason why I added these libraries for typo-js-ts locally. This can be changed anytime, and fix any `false positive` issues.
I tried it locally to produce some false positives, but I couldn't.

eddiejaoude reviewed (approved):
Thank you for the contribution 👍 It looks good 🚀 

Personally I prefer to have the test files in the `tests` directory, but this can always be moved later
